File: tests/maven/java/src/main/java/dev/morphia/critter/test/Person.java
package dev.morphia.critter.test;

import dev.morphia.annotations.CappedAt;
import dev.morphia.annotations.Entity;
import dev.morphia.annotations.Field;
import dev.morphia.annotations.Id;
import dev.morphia.annotations.Index;
import dev.morphia.annotations.Indexes;
import org.bson.types.ObjectId;

import java.util.Objects;
import java.util.StringJoiner;

@Entity(cap = @CappedAt(count = 12))
@Indexes({@Index(fields = @Field("1")), @Index(fields = @Field("2")), @Index(fields = @Field("3"))})
public class Person extends AbstractPerson {
    @Id
    private ObjectId id;

    private String firstName;

    private String lastName;

    private String ssn;

    public Person() {
    }

    public Person(String firstName, String lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
    }

    public ObjectId getId() {
        return id;
    }

    public void setId(final ObjectId id) {
        this.id = id;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public String getSsn() {
        return ssn;
    }

    public void setSsn(String ssn) {
        this.ssn = ssn;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Person)) {
            return false;
        }
        Person person = (Person) o;
        return Objects.equals(id, person.id) && Objects.equals(firstName, person.firstName) && Objects.equals(lastName, person.lastName);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, firstName, lastName);
    }

    @Override
    public String toString() {
        return new StringJoiner(", ", Person.class.getSimpleName() + "[", "]").add("age=" + getAge())
                                                                              .add("id=" + id)
                                                                              .add("firstName='" + firstName + "'")
                                                                              .add("lastName='" + lastName + "'")
                                                                              .toString();
    }
}

File: tests/maven/kotlin/src/main/kotlin/dev/morphia/critter/test/Person.kt
package dev.morphia.critter.test

import com.mongodb.client.model.CollationAlternate.SHIFTED
import dev.morphia.annotations.Collation
import dev.morphia.annotations.Entity
import dev.morphia.annotations.Field
import dev.morphia.annotations.Id
import dev.morphia.annotations.Index
import dev.morphia.annotations.IndexOptions
import dev.morphia.annotations.Indexes
import dev.morphia.annotations.Property
import java.time.LocalDateTime
import kotlin.jvm.internal.ClassReference
import org.bson.types.ObjectId

@Entity
@Indexes(
    Index(fields = arrayOf(Field("emailAddress")), options = IndexOptions(unique = true, collation = Collation(alternate = SHIFTED))),
    Index(fields = arrayOf(Field("ircName"), Field("hostName")))
)
open class Person : AbstractPerson {
    @Id
    var id: ObjectId? = null

    @Property("f")
    var first: String? = null

    var last: String? = null

    var ssn: SSN? = null

    @Suppress("unused")
    constructor()

    constructor(first: String?, last: String?) {
        this.first = first
        this.last = last
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false

        other as Person

        if (id != other.id) return false
        if (first != other.first) return false
        if (last != other.last) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id?.hashCode() ?: 0
        result = 31 * result + (first?.hashCode() ?: 0)
        result = 31 * result + (last?.hashCode() ?: 0)
        return result
    }
    override fun toString(): String {
        return "Person(id=$id, first=$first, last=$last, ssn=$ssn)"
    }
}

@Entity
class SSN(var value: String) {
    private constructor() : this("")
}
